sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -480,7 +480,7 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
                 activeProducer = PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
             }
 
-            using DiagnosticScope scope = CreateDiagnosticScope();
+            using DiagnosticScope scope = CreateDiagnosticScope(eventBatch.AsEnumerable<EventData>());
```

Comment:
I think we should consider pulling out the diagnostic identifier and storing it as part of the batch when the event is added via `TryAdd`, since we're doing the work of instrumenting it and measuring the size at that point.   Something like:

```csharp
public sealed class EventDataBatch : IDisposable
{
        private List<string> EventDiagnosticIdentifiers { get; } = new List<string>();

        // ...

        public bool TryAdd(EventData eventData)
        {
            bool instrumented = EventDataInstrumentation.InstrumentEvent(eventData, FullyQualifiedNamespace, EventHubName);
            bool added = InnerBatch.TryAdd(eventData);

            if (!added && instrumented)
            {
                  EventDataInstrumentation.ResetEvent(eventData);
            }
            else if (EventDataInstrumentation.TryExtractDiagnosticId(eventData, out string diagnosticId))
            {
                 EventDiagnosticIdentifiers.Add(diagnosticId)
            }

            return added;
        }

        internal IReadOnlyList<string> GetEventDiagnosticIdentifiers() => EventDiagnosticIdentifiers;
}
```

At some point, I'm going to get time to fix the EventBatch again to where it returns the `AmqpMessage` which I had to defer when we started to get some weird behavior.